### PR TITLE
🚀 1단계 - 문자열 덧셈 계산기

### PR DESCRIPTION
**Pull request by jivebreaddev**
안녕하세요! 민영님! 

1. 질문사항
- Extractor에 커스텀 인풋이 들어올때 vs 기본 인풋이 들어올 때, 객체 지향적으로 풀 수 있을것같은데 조언이 필요합니다!
- 객체 지향적으로 냄새가 나는 부분은 말씀주시면 고치도록 하겠습니다!

2. 기본 구현 설명
- Pattern은 컴파일해서 사용했습니다. (Effective Java # 6 불필요한 객체 생성을 피하라)
- Calculator는 외부 노출 interface, Extractor가 인풋을 파싱하는 책임을 가집니다.

image

### CODE REVIEW

File: CALCULATOR.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# 1 STEP 요구사항
```
Comment (Gomding) [inline CALCULATOR.md:1]: 요구사항 꼼꼼하게 정리해주셨군요 👍👍👍

File: src/test/java/calculator/NumberExtractor.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  private String extractDelimiter(final String text) {
+    String removePrefix = DELIMITER_PREFIX_REGEX.matcher(text).replaceAll("");
+    String delimiter = DELIMITER_POSTFIX_REGEX.matcher(removePrefix).replaceAll("");
+
+    return delimiter;
+  }
+
+  private void validateEmpty(final String text) {
+    if (ObjectUtils.isEmpty(text))
+      throw new IllegalArgumentException("text is empty: Please check nullity of text");
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: 다른 예외에서는 한글로 예외 메세지를 적어주셨는데
여기는 영어로 적어주셨네요!

통일성을 위해 한 가지로 맞춰주시면 좋겠어요 🤗

File: src/test/java/calculator/NumberExtractor.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  private Integer parseInt(final String text) {
+    try {
+      Integer value = Integer.valueOf(text);
+
+      if (value.compareTo(0) < 0)
+        throw new RuntimeException("숫자가 음수 입니다.");
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: 더 자세한 예외 메세지가 작성되면 좋겠어요! 😊 

사용자에게 현재 상황이 왜 문제인지 자세히 알려주고 
스스로 문제를 해결할 수 있게 작성해보죠!

> 연산 대상인 숫자는 0이상의 양수여야 합니다. 입력된 숫자: -6

File: src/test/java/calculator/NumberExtractor.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    return delimiter;
+  }
+
+  private void validateEmpty(final String text) {
+    if (ObjectUtils.isEmpty(text))
+      throw new IllegalArgumentException("text is empty: Please check nullity of text");
+  }
+
+  private Integer parseInt(final String text) {
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: 여기 private 메서드의 인자는 정말 text 가 맞을까요? 🤔 
text 에서 추출해서 나온 무언가를 인자로 받고있을거라 생각해서요!

인자가 무엇을 의미하는지 표현되면 좋겠어요~!

Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: parseInt 는 총 2가지 일을 하고있다고 생각해요

- 문자열을 숫자로 변경하는 일
- 숫자가 양수인지 검사하는 일

메서드가 하나의 일만 수행하도록 변경해보면 좋겠어요!

+) 숫자라는 값을 포장한 객체로 표현해보는건 어떻게 생각하시나요?
문자열 덧셈 계산기는 양수만을 다루고 있으니 양수라는 객체가 있으면 좋겠군요 🤗

```java
public class 양수 {
    private final int value;
}
```

File: src/test/java/calculator/NumberExtractor.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+      String[] numbers = text
+              .replaceAll(DELIMITER_PREFIX, "")
+              .replaceAll(DELIMITER_POSTFIX, "")
+              .split(delimiter);
+
+      return Arrays.stream(numbers)
+              .filter(StringUtils::hasText)
+              .map(this::parseInt)
+              .toList();
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: if 문 내부의 로직도 private 메서드로 추출해보면 가독성에서 유리할 것 같아요 :)

File: src/test/java/calculator/NumberExtractor.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.List;
+import java.util.regex.Pattern;
+
+class NumberExtractor {
+  private static final String DELIMITER_PREFIX = "//";
+  private static final String DELIMITER_POSTFIX = "\n";
+
+  private static final Pattern DEFAULT_DELIMITER = Pattern.compile("[,:]");
+  private static final Pattern DELIMITER_PREFIX_REGEX = Pattern.compile("[//]");
+  private static final Pattern DELIMITER_POSTFIX_REGEX = Pattern.compile("[\n].*");
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java:16]: 비싼 객체인 Pattern 을 재사용 가능하게 상수로 만들어주신것은 너무 좋네요! 💯 💯 💯

File: src/test/java/calculator/NumberExtractor.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
+package calculator;
+
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+class NumberExtractor {
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java:10]: NumberExtractor 로 책임을 분리해주셨으니 
따로 테스트가 작성되면 좋겠어요 :)

File: src/test/java/calculator/NumberExtractor.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    if (startsWithCustomDelimiter(text)) {
+      String delimiter = extractDelimiter(text);
+
+      String[] numbers = text
+              .replaceAll(DELIMITER_PREFIX, "")
+              .replaceAll(DELIMITER_POSTFIX, "")
+              .split(delimiter);
+
+      return Arrays.stream(numbers)
+              .filter(StringUtils::hasText)
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: hasText 로 필터하는 이유가 궁금해요!

File: src/test/java/calculator/NumberExtractor.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+class NumberExtractor {
+  private static final String DELIMITER_PREFIX = "//";
+  private static final String DELIMITER_POSTFIX = "\n";
+
+  private static final Pattern DEFAULT_DELIMITER = Pattern.compile("[,:]");
+  private static final Pattern DELIMITER_PREFIX_REGEX = Pattern.compile("[//]");
+  private static final Pattern DELIMITER_POSTFIX_REGEX = Pattern.compile("[\n].*");
+
+  // TODO 이 부분 에서 if를 없애는 방향으로 객체지향을 사용할 수 있을 것 같습니다. 좋은 아이디어가 있으시면 공유 부탁드릴께요!
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java]: 추상화를 통해 구현할 수 있을것같긴한데
그만큼 복잡한 구조가 필요한가 라는 생각이 들어요 ㅎㅎ..!

제가 생각했던것은 아래와 같아요!

```java
public interface DelimeterPattern {
    boolean isSupport(String text);

    String[] extractNumbers(String text);
}

public class CustomDelimeterPattern implements DelimeterPattern {
    private static final String DELIMITER_PREFIX = "//";
    private static final String DELIMITER_POSTFIX = "\n";
    private static final Pattern DELIMITER_PREFIX_REGEX = Pattern.compile("[//]");
    private static final Pattern DELIMITER_POSTFIX_REGEX = Pattern.compile("[\n].*");

    @Override
    public boolean isSupport(String text) {
        return this.startsWithCustomDelimiter(text);
    } 

    // .. 생략
}

class NumberExtract {
    List<DelimeterPattern> delimeterPatterns = List.of(new CustomDelimeterPattern(), new DefaultDelimeterPattern());

    public List<Integer> extract(String text) {
        final DelimeterPattern delimeterPattern = delimeterPatterns.stream()
                .filter(it -> it.isSupport())
                .findFirst()
                .orElseThrow(() -> new RuntimeException());
    }
}

```

**Review by Gomding — CHANGES_REQUESTED** [review]
안녕하세요 시창님!
깔끔하게 구현해주셨네요!

몇가지 코멘트 남겼으니 확인부탁드려요 🙏

File: src/test/java/calculator/NumberExtractor.java (hunk 9 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: 디스패쳐서블릿에서 본 코드네요! 

strategy 패턴도 그대로 넣을 수 있을것같네요!

좋은 의견 감사합니다!! 

또, 다른 상황에서 복잡해질 때를 대비해서 한번 같이 의견을 나눠보고 싶었습니다!

File: src/test/java/calculator/NumberExtractor.java (hunk 6 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java:16]: 이펙티브 자바 읽은 보람이 나네요!
칭찬 감사해요

File: src/test/java/calculator/NumberExtractor.java (hunk 4 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: `private Integer parseInt(final String element) {
    try {
      Integer value = Integer.valueOf(element);`

로 element로 표현하였습니다!

File: src/test/java/calculator/NumberExtractor.java (hunk 3 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: `String.format("연산 대상의 숫자는 0 이상의 양수여야 합니다. 입력된 숫자: %d", value)`

다음과 같은 값으로 자세하게 변경하였습니다.

File: src/test/java/calculator/NumberExtractor.java (hunk 2 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: `  throw new IllegalArgumentException("변환 대상 인풋이 빈값이거나 null 입니다.");`
다음과 같이 변경하였습니다.

File: src/test/java/calculator/NumberExtractor.java (hunk 8 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: list내에 빈값이 들어갈 경우 처리를 하려고 필터했습니다.

나름의 edge case로 진행했는데 제거하도록 하겠습니다.

File: src/test/java/calculator/NumberExtractor.java (hunk 4 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java]: 해당 부분도 숫자가 양수인지 검사하는 일을 값을 포장한 객체로 표현하는 것을 고려했었는데 
다음과 같은 이유에서 배제 했었습니다.


1. Wrapper class 대신 기본자료형을 사용하라 # 49 Effective Java
참조: https://kimss1502.github.io/java/49.-Wrapper-class-%EB%8C%80%EC%8B%A0-%EA%B8%B0%EB%B3%B8%EC%9E%90%EB%A3%8C%ED%98%95%EC%9D%84-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC/

하지만 추후에 로직이 복잡해진다면, 민영님이 말씀하신 로직 검증의 책임을 도메인이 하는것이 필요해 보입니다!
수정 진행하겠습니다!

**Comment by jivebreaddev** [comment]
수정 요청사항 종합

1. 오류 메시지 구체화
2. private method 내 가독성을 위해 메서드 분리
3. hasText 필터 제거
4. NumberExtractor.parseInt의 책임 중 양수 검증은 도메인 객체로 이전
5. NumberExtractor 에 대한 test 도입


image

File: src/test/java/calculator/PositiveNumber.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
+package calculator;
+
+import java.util.Objects;
+
+class PositiveNumber {
```
Comment (Gomding) [inline src/test/java/calculator/PositiveNumber.java:5]: 덧셈 계산기의 숫자를 포장한 객체 👍

File: src/test/java/calculator/NumberExtractor.java [hunk 11]
```diff
@@ -54,21 +62,17 @@ private String extractDelimiter(final String text) {
...
-      if (value.compareTo(0) < 0)
-        throw new RuntimeException("숫자가 음수 입니다.");
 
-      return value;
+      return PositiveNumber.from(element);
 
     } catch (NumberFormatException e) {
 
-      throw new RuntimeException("숫자가 아닌 값이 text에 있습니다.");
+      throw new RuntimeException(String.format("연산 대상은 숫자이어야 합니다. 입력된 글자: %s", element));
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java:75]: 숫자인지 아닌지 검사도 PositiveNumber 객체에게 맞겨보는건 어떠신가요!

PositiveNumber 객체 생성 시 NumberFormatException 을 처리해보는 것이죠 :)
팩터리 메서드(PositiveNumber.from)에서 문자열을 받고있으니 숫자인지 검사와 처리도 함께 해줘야한다고 생각해요!

File: src/test/java/calculator/PositiveNumber.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  private final Integer value;
```
Comment (Gomding) [inline src/test/java/calculator/PositiveNumber.java:6]: 한 가지 궁금한 건 
원시 타입 int 대신 Integer 를 사용하신 이유가 궁금했어요 🙋

File: src/test/java/calculator/NumberExtractor.java [hunk 13]
```diff
@@ -15,27 +15,35 @@ class NumberExtractor {
...
+    String[] numbers = text
+            .replaceAll(DELIMITER_PREFIX, "")
+            .replaceFirst(delimiter, "")
+            .replaceAll(DELIMITER_POSTFIX, "")
+            .split(delimiter);
+    return numbers;
+  }
+
+  @NotNull
+  private List<PositiveNumber> extractIntegers(String[] numbers) {
```
Comment (Gomding) [inline src/test/java/calculator/NumberExtractor.java:46]: PositiveNumber 의 컬렉션을 일급컬렉션 객체로 만들어봐도 좋습니다 :)

File: src/test/java/calculator/PositiveNumber.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    return new PositiveNumber(value);
+  }
+
+  public Integer getValue() {
+    return value;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (Gomding) [inline src/test/java/calculator/PositiveNumber.java:26]: equals & hashcode 도 꼼꼼하게 재정의해주셨군요 👍 👍 👍 
양수 객체이니 내부 값으로 동등성 비교 😆

File: src/test/java/calculator/StringCalculator.java [hunk 15]
```diff
@@ -24,9 +23,10 @@ public int add(final String text) {
       return ZERO;
     }
 
-    List<Integer> numbers = extractor.extract(text);
+    List<PositiveNumber> numbers = extractor.extract(text);
 
     return numbers.stream()
+            .map(PositiveNumber::getValue)
             .reduce(0, Integer::sum);
```
Comment (Gomding) [inline src/test/java/calculator/StringCalculator.java:30]: PositiveNumber 에 sum 이라는 메서드를 만들어봐도 좋겠어요 :)

**Review by Gomding — APPROVED** [review]
안녕하세요 시창님 🙌
코멘트 잘 반영해주셨군요!
이미 깔끔하게 구현해주셔서 다음단계 진행을 위해 빠르게 merge 하겠습니다 :)

다음 단계에서 뵙겠습니다 🙇‍♂️

File: src/test/java/calculator/NumberExtractor.java (hunk 11 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java:75]: 생각이 거기까지 미치지 못했네요!

1. wrapper 객체가 string -> int 로 변환하면서 숫자인지 + 숫자가 아닌지 검사 처리 하는 책임 가진다.

Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java:75]: 정말 좋은 리뷰인것같습니다!

File: src/test/java/calculator/NumberExtractor.java (hunk 13 above)
Comment (jivebreaddev) [reply src/test/java/calculator/NumberExtractor.java:46]: 일급 컬렉션으로 만들어서 sum 을 구현하면 도메인 로직  까지 전부 깔끔하게 묶이겠네요!

File: src/test/java/calculator/PositiveNumber.java (hunk 12 above)
Comment (jivebreaddev) [reply src/test/java/calculator/PositiveNumber.java:6]: 원시타입과 wrapper 타입을 사용할때, 
 `오토박싱과 언박싱에 대해서 성능적 영향도가 생길 수 있지 않을까?` 라는 생각이 컸고
 `jpa 에서도 persistence entity와의 연산`에서 보통 Wrapper를 사용하게 되니깐 습관적으로 사용하게 된것 같습니다.

보통 원시값을 사용하시나요?
